From 743520f67cf900ca637dd86e4e16e20185760834 Mon Sep 17 00:00:00 2001
From: Rico Ganahl <rico.ganahl@bytesatwork.ch>
Date: Mon, 22 Apr 2024 11:37:31 +0200
Subject: [PATCH] bytesatwork: imx8mp: prepare dynamic ram config

Use bawconfig tool to dynamically program the assembled RAM type.

Signed-off-by: Rico Ganahl <rico.ganahl@bytesatwork.ch>
---
 MAINTAINERS                                   |  1 +
 .../bytedevkit_imx8mp/imx8mp_bytedevkit.c     | 24 ++++++++++++
 .../bytedevkit_imx8mp/lpddr4_timing.c         |  3 +-
 .../bytedevkit_imx8mp/lpddr4_timings.h        | 12 ++++++
 board/bytesatwork/bytedevkit_imx8mp/spl.c     | 38 +++++++++++++++++++
 include/configs/imx8mp_bytedevkit.h           |  1 +
 6 files changed, 78 insertions(+), 1 deletion(-)
 create mode 100644 board/bytesatwork/bytedevkit_imx8mp/lpddr4_timings.h

diff --git a/MAINTAINERS b/MAINTAINERS
index 1572bff745..44129f4d56 100644
--- a/MAINTAINERS
+++ b/MAINTAINERS
@@ -212,6 +212,7 @@ T:	git https://github.com/bytesatwork/u-boot-imx
 F:	arch/arm/dts/imx8mp-bytedevkit-u-boot.dtsi
 F:	arch/arm/dts/imx8mp-bytedevkit.dts
 F:	board/bytesatwork/bytedevkit_imx8mp/
+F:	bytesatwork/bytedevkit_imx8mp/lpddr4_timings.h
 F:	cmd/cpuinfo.c
 F:	configs/imx8mp_bytedevkit_defconfig
 F:	include/configs/imx8mp_bytedevkit.h
diff --git a/board/bytesatwork/bytedevkit_imx8mp/imx8mp_bytedevkit.c b/board/bytesatwork/bytedevkit_imx8mp/imx8mp_bytedevkit.c
index fcc41b83e6..5fc933ddf2 100644
--- a/board/bytesatwork/bytedevkit_imx8mp/imx8mp_bytedevkit.c
+++ b/board/bytesatwork/bytedevkit_imx8mp/imx8mp_bytedevkit.c
@@ -541,3 +541,27 @@ int is_recovery_key_pressing(void)
 }
 #endif /* CONFIG_ANDROID_RECOVERY */
 #endif /* CONFIG_FSL_FASTBOOT */
+
+#include <asm/armv8/mmu.h>
+extern struct mm_region *mem_map;
+
+int board_phys_sdram_size(phys_size_t *size)
+{
+	/* NOTE: value passed from SPL (same address as set in SPL) */
+	phys_size_t *ram_size = (phys_size_t *)BAW_CONFIG_RAM_SIZE_PTR_ADDR;
+
+	if (!size)
+		return -EINVAL;
+
+	switch (*ram_size) {
+	case SZ_4G:
+		mem_map[6].size = SZ_2G + SZ_1G;
+		mem_map[7].size = SZ_1G;
+		*size = *ram_size;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
diff --git a/board/bytesatwork/bytedevkit_imx8mp/lpddr4_timing.c b/board/bytesatwork/bytedevkit_imx8mp/lpddr4_timing.c
index bc43c0a38b..e9135dd5eb 100644
--- a/board/bytesatwork/bytedevkit_imx8mp/lpddr4_timing.c
+++ b/board/bytesatwork/bytedevkit_imx8mp/lpddr4_timing.c
@@ -9,6 +9,7 @@
 
 #include <linux/kernel.h>
 #include <asm/arch/ddr.h>
+#include "lpddr4_timings.h"
 
 struct dram_cfg_param ddr_ddrc_cfg[] = {
 	/** Initialize DDRC registers **/
@@ -1831,7 +1832,7 @@ struct dram_fsp_msg ddr_dram_fsp_msg[] = {
 };
 
 /* ddr timing config params */
-struct dram_timing_info dram_timing = {
+struct dram_timing_info dram_timing_4gb = {
 	.ddrc_cfg = ddr_ddrc_cfg,
 	.ddrc_cfg_num = ARRAY_SIZE(ddr_ddrc_cfg),
 	.ddrphy_cfg = ddr_ddrphy_cfg,
diff --git a/board/bytesatwork/bytedevkit_imx8mp/lpddr4_timings.h b/board/bytesatwork/bytedevkit_imx8mp/lpddr4_timings.h
new file mode 100644
index 0000000000..40a5fa9ded
--- /dev/null
+++ b/board/bytesatwork/bytedevkit_imx8mp/lpddr4_timings.h
@@ -0,0 +1,12 @@
+/* SPDX-License-Identifier: GPL-2.0+ */
+/*
+ * Copyright (C) 2024 bytesatwork AG - https://www.bytesatwork.io
+ */
+
+#ifndef _LPDDR4_TIMINGS_H_
+#define _LPDDR4_TIMINGS_H_
+
+/* NOTE: generated timings for different RAM types. */
+extern struct dram_timing_info dram_timing_4gb;
+
+#endif /* _LPDDR4_TIMINGS_H_ */
diff --git a/board/bytesatwork/bytedevkit_imx8mp/spl.c b/board/bytesatwork/bytedevkit_imx8mp/spl.c
index a05441cfe6..9e34b9ed77 100644
--- a/board/bytesatwork/bytedevkit_imx8mp/spl.c
+++ b/board/bytesatwork/bytedevkit_imx8mp/spl.c
@@ -29,7 +29,10 @@
 #include <fsl_esdhc_imx.h>
 #include <mmc.h>
 #include <asm/arch/ddr.h>
+
+#include "lpddr4_timings.h"
 #include "../baw_config/baw_config_eeprom.h"
+#include "../baw_config/baw_config_get.h"
 
 DECLARE_GLOBAL_DATA_PTR;
 
@@ -59,12 +62,47 @@ int spl_board_boot_device(enum boot_device boot_dev_spl)
 #endif
 }
 
+/* NOTE: ddr timing config params, see arch/arm/include/asm/arch-imx8m/ddr.h */
+struct dram_timing_info dram_timing;
+
+struct dram_setup {
+	enum baw_config_ram ram;
+	const phys_size_t size;
+	const struct dram_timing_info *timing;
+};
+
 void spl_dram_init(void)
 {
+	/* NOTE: store ram_size in OCRAM to pass value from SPL to u-boot */
+	phys_size_t *ram_size = (phys_size_t *)BAW_CONFIG_RAM_SIZE_PTR_ADDR;
+	const struct dram_setup dram_setups[] = {
+		{M7_RAM_MT53E1G32D2FW_046, SZ_4G, &dram_timing_4gb},
+		{0, 0, NULL },
+	};
+	struct baw_config config;
+	bool ram_found = false;
+
 	if (IS_ENABLED(CONFIG_BAW_CONFIG_EEPROM))
 		if (baw_config_eeprom_init())
 			printf("EEPROM init failed!\n");
 
+	baw_config_get(&config);
+	for (int i = 0; dram_setups[i].ram > 0; i++) {
+		if (config.ram == dram_setups[i].ram) {
+			memcpy(&dram_timing, dram_setups[i].timing,
+			       sizeof(dram_timing));
+			*ram_size = dram_setups[i].size;
+			ram_found = true;
+			break;
+		}
+	}
+
+	if (!ram_found) {
+		printf("Error: Unknown RAM config, use fallback.\n");
+		memcpy(&dram_timing, dram_setups[0].timing, sizeof(dram_timing));
+		*ram_size = dram_setups[0].size;
+	}
+
 	ddr_init(&dram_timing);
 }
 
diff --git a/include/configs/imx8mp_bytedevkit.h b/include/configs/imx8mp_bytedevkit.h
index ed876e71c9..ac8b083112 100644
--- a/include/configs/imx8mp_bytedevkit.h
+++ b/include/configs/imx8mp_bytedevkit.h
@@ -18,6 +18,7 @@
 #define BAW_CONFIG_BUILTIN_PCB		M7_PCB_REV_1_0
 #define BAW_CONFIG_BUILTIN_RAM		M7_RAM_MT53E1G32D2FW_046
 #define BAW_CONFIG_BUILTIN_FLASH	EMMC_16GB
+#define BAW_CONFIG_RAM_SIZE_PTR_ADDR	0x98ffc0
 
 #define CONFIG_SYS_BOOTM_LEN		(32 * SZ_1M)
 
-- 
2.39.2

